File: leetcode/problems/easy/383.ransom-note/python/solution.py
# https://leetcode.com/problems/ransom-note/description/
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        char_count = {}
        
        for char in magazine:
            char_count[char] = char_count.get(char, 0) + 1
        
        for char in ransomNote:
            if char not in char_count or char_count[char] <= 0:
                return False
            char_count[char] -= 1
        
        return True

# 지렸다
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        
        for i in ransomNote:
            if i in magazine:
                magazine = magazine.replace(i,"",1)
            else: return False
        
        return True

File: leetcode/problems/easy/383.ransom-note/kotlin/Solution.kt
// https://leetcode.com/problems/ransom-note/?envType=problem-list-v2&envId=rab78cw1
class Solution {
    fun canConstruct(ransomNote: String, magazine: String): Boolean {
        val letters = IntArray(26)
        for (char in ransomNote) {
            letters[char - 'a']++
        }

        for (char in magazine) {
            letters[char - 'a']--
        }

        return letters.all { it <= 0 }
    }

    fun canConstruct2(ransomNote: String, magazine: String): Boolean {
        val letters = IntArray(26)
        var magazineIndex = 0
        for (char in ransomNote) {
            val letterIndex = char - 'a'
            while(letters[letterIndex] == 0) {
                if (magazineIndex >= magazine.length) return false
                letters[magazine[magazineIndex] - 'a']++
                magazineIndex++
            }
            letters[letterIndex]--
        }


        return true
    }   
}
